plugins/org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/actions/ImportsAwareClipboardAction.java

```diff
@@ -181,47 +180,65 @@ public XbaseClipboardData apply(Clipboard input) {
 				});
 		JavaImportData javaImportsContent = ClipboardUtil.getJavaImportsContent();
 		String textFromClipboard = ClipboardUtil.getTextFromClipboard();
+		boolean addImports = shouldAddImports(EditorUtils.getXtextEditor(getTextEditor()));
 		if (xbaseClipboardData != null && !sameTarget(xbaseClipboardData)) {
-			doPasteXbaseCode(xbaseClipboardData);
-		} else if (javaImportsContent != null && canPasteJava(textFromClipboard, getJavaCodeConverter())) {
-			doPasteJavaCode(textFromClipboard, javaImportsContent, getJavaCodeConverter());
+			doPasteXbaseCode(xbaseClipboardData, addImports);
+		} else if (javaImportsContent != null) {
+			doPasteJavaCode(textFromClipboard, javaImportsContent, addImports);
 		} else {
 			textOperationTarget.doOperation(operationCode);
 		}
 	}
 
-	private boolean canPasteJava(String javaToConvert, IJavaCodeConverter javaCodeConverter) {
-		if (javaCodeConverter == null) {
-			return true;
+	/**
+	 * Should not add imports when pasting into a {@link XStringLiteral} or Comments (except of JavaDoc)
+	 * 
+	 * @param xtextEditor
+	 */
+	private boolean shouldAddImports(XtextEditor xtextEditor) {
+		int caretOffset = caretOffset(xtextEditor);
+		String typeRight = IDocument.DEFAULT_CONTENT_TYPE;
+		String typeLeft = IDocument.DEFAULT_CONTENT_TYPE;
+		try {
+			typeRight = TextUtilities.getContentType(xtextEditor.getDocument(),
+					IDocumentExtension3.DEFAULT_PARTITIONING, caretOffset, false);
+			typeLeft = TextUtilities.getContentType(xtextEditor.getDocument(), IDocumentExtension3.DEFAULT_PARTITIONING,
+					caretOffset > 0 ? caretOffset - 1 : caretOffset, false);
+		} catch (BadLocationException exception) {
+			// Should not happen
 		}
-		XtextEditor xtextEditor = EditorUtils.getXtextEditor(getTextEditor());
-		EObject targetElement = xtextEditor.getDocument().priorityReadOnly(new IUnitOfWork<EObject, XtextResource>() {
-
-			@Override
-			public EObject exec(XtextResource state) throws Exception {
-				EObject elementAtOffset = helper.resolveContainedElementAt(state, 10);
-				return elementAtOffset;
-			}
-		});
+		if (COMMENT_PARTITION.equals(typeRight) || STRING_LITERAL_PARTITION.equals(typeRight)
+				|| SL_COMMENT_PARTITION.equals(typeRight) || "__rich_string".equals(typeRight)) {
```

Comment:
Should we have a service / an extension interface for the partition type mapper that returns the list of partitions where imports shouldn't be added? Here we add a dep on Xtends partition type in xbase